File: java-implementation/src/main/java/com/inventory/products/guitars/Electric.java
package com.inventory.products.guitars;

// Represents an electric guitar as a concrete subclass of Guitar.
public class Electric extends Guitars {
    private String pickupType; // Type of pickup used in the electric guitar

    // Constructs an ElectricGuitar instance with specified details
    public Electric(int id, String name, double price, int volume, String material, int stringCount, String pickupType, String supplier) {
        super(id, name, price, volume, material, stringCount, supplier);
        this.pickupType = pickupType;
    }

    // Calculates the discount based on volume and month
    @Override
    public double calculateDiscount(int volume, int month) {
        if (month == 12) {
            return price * volume * 0.50;
        }
        return 0;
    }

    // Provides a string representation of the electric guitar's details
    @Override
    public String displayProductInfo() {
        return "ID: " + id + "\nElectric Guitar: " + name + "\nMaterial: " + material
               + "\nPickup Type: " + pickupType + "\nPrice: €" + price
               + "\nString Count: " + stringCount + "\nVolume: " + volume + " units\nSupplier: " + supplier + "\n\n";
    }
}

File: cpp-implementation/src/products/guitars/Electric.cpp
#include "Product.ih"

// Constructor implementation for electric guitars
Electric::Electric(int id, string name, double price, int volume, 
                string material, int stringCount, 
                 string pickupType, string supplier)
    : Guitars(id, name, price, volume, material, stringCount, supplier), 
      pickupType(pickupType) {}

// Calculates the discount depending on the current month and the packaged quantity
double Electric::calculateDiscount(int volume, int month) const {
    if (month == 12) {
        return price * volume * 0.50;
    }
    return 0;
}

// Displays all the information 
string Electric::displayProductInfo() const {
    ostringstream info;
    info << "ID: " << id << "\nElectric Guitar: " << name << "\nSupplier: " << getSupplier()
         << "\nMaterial: " << material << "\nPickup Type: " << pickupType
         << "\nPrice: €" << price << "\nString Count: " << stringCount
         << "\nVolume: " << volume << " units\n\n";
    return info.str();
}

